File: pyfiles/figurepro.py
from functools import singledispatch
import os
from matplotlib.pyplot import *
from matplotlib.transforms import Bbox                                                                         
import re

class FigurePro(Figure):
   ax=None
   ax2=[]
   desc=""
   label=""
   ww=6.0
   hh=6.0/1.618
   rect=None
   markedLine={}
   def __init__(self,
                figsize=None,
                dpi=None,
                facecolor=None,
                edgecolor=None,
                linewidth=0.0,
                frameon=None,
                subplotpars=None,  # rc figure.subplot.*
                tight_layout=None,  # rc figure.autolayout
                constrained_layout=None,  # rc figure.constrained_layout.use
                rect=(0.2,0.2,0.7,0.7), ## NEW ARGUMENTS
                ww=6.0,hh=4.5,
                desc="",
                label=""):
      super().__init__(figsize, dpi, facecolor, edgecolor, linewidth,
                frameon, subplotpars, tight_layout, constrained_layout)

      self.rect=rect
      self.ww=ww
      self.hh=hh
      self.set_size_inches(ww,hh)
      #setattr(self,'ax',self.add_axes(rect, label="main"))
      self.ax=self.add_axes(rect, label="main")
      self.ax.patch.set_alpha(0.0)
      self.desc = desc
      self.label =label 
      self.ax2=[]
      self.markedLine={}

   def adddesc(self,atext,newline=True):
      self.desc="%s\n%s" % (self.desc, atext) 

   def mirrorax(self):
      n=len(self.ax2)
      #self.ax2.append(self.add_axes(self.rect, label="%d" % (n,)))
      self.ax2.append(self.ax.twinx())
      self.ax2[-1].patch.set_alpha(0.0)
   def mirroray(self):
      n=len(self.ax2)
      #self.ax2.append(self.add_axes(self.rect, label="%d" % (n,)))
      self.ax2.append(self.ax.twiny())
      self.ax2[-1].patch.set_alpha(0.0)

   def pgfprint(self,odir='.',frmt='eps',tbm='tbmb',auxfrmt='png',auxleg=True, 
      suffix="",
      README=None, panel=None):

      if len(self.label)>0 and panel:
         figlabel="%s_%s" % (self.label, panel)
      elif len(self.label)>0:
         figlabel=self.label
      elif panel:
         figlabel=panel
      else:
         figlabel=""
      if suffix:
         if figlabel:
            figlabel=figlabel+"_"+suffix
         else:
            figlabel=suffix
     
      # save whole figure (for safety purposes)
      if len(figlabel)>0:
         ofnm_fig="%s/fig_%d-%s_WITHAXES.%s"  % (odir,self.number,figlabel,auxfrmt)
      else:
         ofnm_fig="%s/fig_%d_WITHAXES.%s"  % (odir,self.number,auxfrmt)
      if auxleg:
         l=self.ax.legend(bbox_to_anchor=(1.0,0.0,1,1))
         self.set_size_inches(4*self.ww,2*self.hh)
         bb=self.ax.get_position()
         self.ax.set_position(Bbox([[0.2,0.2],[0.5,0.8]]))
         for i1 in range(len(self.ax2)):
            l=self.ax2[i1].legend(bbox_to_anchor=(1.0,0.0,1,0.5-i1*0.2))
            #self.set_size_inches(4*self.ww,2*self.hh)
            bb=self.ax2[i1].get_position()
            self.ax2[i1].set_position(Bbox([[0.2,0.2],[0.5,0.8]]))

      self.savefig(ofnm_fig)
      if auxleg:
         l.remove()
         self.set_size_inches(self.ww,self.hh)

      # save clean axes
      if len(figlabel)>0:
         ofnm_fig="%s/fig_%d-%s.%s"  % (odir,self.number,figlabel,frmt)
      else:
         ofnm_fig="%s/fig_%d.%s"  % (odir,self.number,frmt)
      bb=self.ax.get_position()
      self.ax.set_position(Bbox([[0.0,0.0],[1.0,1.0]]))
      self.ax.axis("off")
      for i1 in range(len(self.ax2)):
         self.ax2[i1].set_position(Bbox([[0.0,0.0],[1.0,1.0]]))
         self.ax2[i1].axis("off")
         
      if frmt=="png":
         self.savefig(ofnm_fig,dpi=400)
      else:
         self.savefig(ofnm_fig)
      self.ax.set_position(bb)
      self.ax.axis("on")
      for i1 in range(len(self.ax2)):
         self.ax2[i1].set_position(bb)
         self.ax2[i1].axis("on")
      
      # save .tex file
      if len(figlabel)>0:
         ofnm_tex="%s/fig_%d-%s.tex"  % (odir,self.number,figlabel)
      else:
         ofnm_tex="%s/fig_%d.tex"  % (odir,self.number)
      xlim=self.ax.get_xlim()
      ylim=self.ax.get_ylim()
      f=open(ofnm_tex,'w')
      # write header as comments
      bb=self.bbox_inches
      f.write("%% width=%f in, \n" % bb.width  )
      f.write("%% height=%f in, \n" % bb.height  )
      f.write("%% xlabel={%s}, \n" % self.ax.get_xlabel()  )
      f.write("%% ylabel={%s}, \n" % self.ax.get_ylabel()  )
      f.write("%%\n")
      f.write("%% Line2D objects: \n")
      alist=self.ax.get_children()
      for item in alist:
         if item.__class__.__name__ == 'Line2D':
            label=item.get_label()
            if len(label)>0 and not label.startswith('_'): 
               f.write("%% %s : %s\n" % (item.get_label(), item.get_color()))
      # write line that can be directly included with \input{file}
      if len(figlabel)>0:
         f.write("\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d-%s.%s};" % (xlim[0],xlim[1],ylim[0],ylim[1],tbm,self.number,figlabel,frmt))
      else:
         f.write("\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d.%s};" % (xlim[0],xlim[1],ylim[0],ylim[1],tbm,self.number,frmt))
      f.close()

      for i1 in range(len(self.ax2)):
         # save .tex file
         if len(figlabel)>0:
            ofnm_tex="%s/fig_%d-%s-extra_%d.tex"  % (odir,self.number,figlabel,i1)
         else:
            ofnm_tex="%s/fig_%d-extra_%d.tex"  % (odir,self.number,i1)
         xlim=self.ax2[i1].get_xlim()
         ylim=self.ax2[i1].get_ylim()
         f=open(ofnm_tex,'w')
         # write header as comments
         bb=self.bbox_inches
         f.write("%% width=%f in, \n" % bb.width  )
         f.write("%% height=%f in, \n" % bb.height  )
         f.write("%% xlabel={%s}, \n" % self.ax2[i1].get_xlabel()  )
         f.write("%% ylabel={%s}, \n" % self.ax2[i1].get_ylabel()  )
         f.write("%%\n")
         f.write("%% Line2D objects: \n")
         alist=self.ax2[i1].get_children()
         for item in alist:
            if item.__class__.__name__ == 'Line2D':
               label=item.get_label()
               if len(label)>0 and not label.startswith('_'): 
                  f.write("%% %s : %s\n" % (item.get_label(), item.get_color()))
         # write line that can be directly included with \input{file}
         if len(figlabel)>0:
            f.write("\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d-%s.%s};" % (xlim[0],xlim[1],ylim[0],ylim[1],tbm,self.number,figlabel,frmt))
         else:
            f.write("\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d.%s};" % (xlim[0],xlim[1],ylim[0],ylim[1],tbm,self.number,frmt))
         f.close()

      if not README==None:
         if figlabel:
            README.write(" Figure %d / %s \n" % (self.number,figlabel))
         else:
            README.write(" Figure %d \n" % self.number)
         README.write(f" ww,hh = {self.ww},{self.hh} \n")
         README.write(f" ww/hh = {self.ww/self.hh} \n")
         README.write("\n")
         README.write(" %s\n" % self.desc)
         README.write("\n")
         README.write(" Line2D items: \n")
         for item in alist:
            if item.__class__.__name__ == 'Line2D':
               label=item.get_label()
               if len(label)>0 and not label.startswith('_'): 
                  README.write("    - %s : %s\n" % (item.get_label(), item.get_color()))
         README.write("\n")
         README.write("\n")
   
   def saveLine2D(self,odir='.',labelList=[],README=None, panel=None):

      if len(self.label)>0:
         odir_data="%s/fig_%d-%s_data"  % (odir,self.number,self.label)
      else:
         odir_data="%s/fig_%d_data"  % (odir,self.number)
      if not os.path.isdir(odir_data): os.system('mkdir %s' % odir_data)

      if panel: odir_data="%s/%s"  % (odir_data,panel)
      if not os.path.isdir(odir_data): os.system('mkdir %s' % odir_data)

      if not README==None:
         README.write(" ASCII data of Line2D objects from Figure %d:\n" % (self.number,))

      alist=self.ax.get_children()
      for item in alist:
         if item.__class__.__name__ == 'Line2D':
            label=item.get_label()
            if len(labelList) > 0 and not label in labelList: continue
            if len(label)>0 and not label.startswith('_'): 
               header="%s - %s" % (item.get_label(), item.get_color())
               ofnm="%s/%s.dat"  % (odir_data, item.get_label().replace(' ','_'))
               if not README==None:
                  README.write(" - %s \n" % (header,))
               f=open(ofnm,'w')
               f.write("%25s %25s\n" % ("x","y")) 
               x=item.get_xdata(orig=False)
               y=item.get_ydata(orig=False)
               n=x.shape; n=n[0]
               for j in range(n):
                   f.write("%25.15E %25.15E\n" % (x[j], y[j])) 
               f.close()
      for i in range(len(self.ax2)):
         alist=self.ax2[i].get_children()
         for item in alist:
            if item.__class__.__name__ == 'Line2D':
               label=item.get_label()
               if len(labelList) > 0 and not label in labelList: continue
               if len(label)>0 and not label.startswith('_'): 
                  header="%s - %s" % (item.get_label(), item.get_color())
                  ofnm="%s/%s.dat"  % (odir_data, item.get_label().replace(' ','_'))
                  if not README==None:
                     README.write(" - %s \n" % (header,))
                  f=open(ofnm,'w')
                  f.write("%25s %25s\n" % ("x","y")) 
                  x=item.get_xdata(orig=False)
                  y=item.get_ydata(orig=False)
                  n=x.shape; n=n[0]
                  for j in range(n):
                      f.write("%25.15E %25.15E\n" % (x[j], y[j])) 
                  f.close()
      if not README==None:
         README.write("\n\n")
  
   def setww(self,ww_):
      self.ww=ww_
      self.set_size_inches(self.ww,self.hh)
   def sethh(self,hh_):
      self.hh=hh_
      self.set_size_inches(self.ww,self.hh)

   def addNMarkers(self,label=None,marker='o',nmark=3,dshift=0,props={},xlim=False):
      localprops=props.copy()
      if not 'marker' in localprops.keys(): localprops['marker']=marker
      if not 'ls' in localprops.keys(): localprops['ls']='none'
      alist=self.ax.get_children()
      for item in alist:
         if item.__class__.__name__ == 'Line2D':
            if label==item.get_label() or (label is None and not item.get_label().startswith('_')):
               newlabel='%s_marked' % (item.get_label(),)
               x=item.get_xdata(orig=False)
               y=item.get_ydata(orig=False)
               if not xlim:
                  n=x.shape; n=n[0]
                  ii=np.linspace(0,n-1,nmark+1,dtype='i')
                  ii=ii[0:-1]
                  ishift=int(dshift*n/nmark)
                  ii=(ii+ishift)%n
               else:
                  xl=self.ax.get_xlim()
                  jj=np.where( ( x>=xl[0] ) & ( x<=xl[1] ) )
                  if len(jj[0])==0:
                     ii=[]
                  else:
                     n=len(jj[0])
                     kk=np.linspace(0,n-1,nmark+1,dtype='i')
                     kk=kk[0:-1]
                     ishift=int(dshift*n/nmark)
                     kk=(kk+ishift)%n
                     ii=jj[0][kk]
               if len(ii)==0:
                  self.markedLine[newlabel]=self.ax.plot([],[],label=newlabel,**localprops)
               else:
                  self.markedLine[newlabel]=self.ax.plot(x[ii],y[ii],label=newlabel,**localprops)
      for i in range(len(self.ax2)):
         alist=self.ax2[i].get_children()
         for item in alist:
            if item.__class__.__name__ == 'Line2D':
               if label==item.get_label() or (label is None and not item.get_label().startswith('_')):
                  newlabel='%s_marked' % (item.get_label(),)
                  x=item.get_xdata(orig=False)
                  y=item.get_ydata(orig=False)
                  if not xlim:
                     n=x.shape; n=n[0]
                     ii=np.linspace(0,n-1,nmark+1,dtype='i')
                     ii=ii[0:-1]
                     ishift=int(dshift*n/nmark)
                     ii=(ii+ishift)%n
                  else:
                     xl=self.ax2[i].get_xlim()
                     jj=np.where( ( x>=xl[0] ) & ( x<=xl[1] ) )
                     if len(jj[0])==0:
                        ii=[]
                     else:
                        n=len(jj[0])
                        kk=np.linspace(0,n-1,nmark+1,dtype='i')
                        kk=kk[0:-1]
                        ishift=int(dshift*n/nmark)
                        kk=(kk+ishift)%n
                        ii=jj[0][kk]
                  if len(ii)==0:
                     self.markedLine[newlabel]=self.ax2[i].plot([],[],label=newlabel,**localprops)
                  else:
                     self.markedLine[newlabel]=self.ax2[i].plot(x[ii],y[ii],label=newlabel,**localprops)

   def removeMarkedLine_all(self):
      keystoremove=[]
      for key,value in self.markedLine.items():
         value.pop(0).remove()
         keystoremove.append(key)
      for key in keystoremove:
         self.markedLine.pop(key)

   def removeMarkedLine_bylabel(self,label):
      newlabel='%s_marked' % (label,)
      item=self.markedLine[newlabel]
      item.pop(0).remove()

   def align_yvalue_around_main(self,val=0.0,iax=0):
      axref = self.ax;
      ax    = self.ax2[iax];
      ylref = axref.get_ylim()
      yl    = ax.get_ylim()
      loold=yl[0] 
      hiold=yl[1] 
      loref=ylref[0] 
      hiref=ylref[1] 
      m=(hiref-val)/(val-loref)
      # assume 
      hi=val+(val-loold)*m
      if hi>hiold: 
         lo=loold
      else:
         hi=hiold
         lo=val-(hiold-val)/m
      ax.set_ylim([lo,hi])  
      return
   def align_xvalue_around_main(self,val=0.0,iax=0):
      axref = self.ax;
      ax    = self.ax2[iax];
      ylref = axref.get_xlim()
      yl    = ax.get_xlim()
      loold=yl[0] 
      hiold=yl[1] 
      loref=ylref[0] 
      hiref=ylref[1] 
      m=(hiref-val)/(val-loref)
      # assume 
      hi=val+(val-loold)*m
      if hi>hiold: 
         lo=loold
      else:
         hi=hiold
         lo=val-(hiold-val)/m
      ax.set_xlim([lo,hi])  
      return

   def highlight_regex(self,patt): 
      self.patt = patt
      p = re.compile(patt)
      alist=self.ax.get_children()
      for item in alist:
         if item.__class__.__name__ == 'Line2D':
            m = p.match(item.get_label())
            if m is None:
               item.oldcolor = item.get_color()
               item.set_color('0.5')
   def highlight_reset(self): 
      p = re.compile(self.patt)
      alist=self.ax.get_children()
      for item in alist:
         if item.__class__.__name__ == 'Line2D':
            m = p.match(item.get_label())
            if m is None:
               item.set_color(item.oldcolor)
         
    

File: mfiles/figurepro.m
classdef figurepro

properties
handle;
label="";
ax;
axlist;
ww=6.0;
hh=6.0/1.618;
desc="";
papercolor='w';
end

methods
% Constructor
function self = figurepro(varargin)

   p=inputParser();
   addOptional(p,'label',self.label);
   parse(p,varargin{:});
   self.label=p.Results.label;

   self.handle=figure();
   set(self.handle,'Units','inches');
   self.axlist(1)=axes('Parent',self.handle);
   self.ax=self.axlist(1);
   hold(self.ax,'on');
   box(self.ax,'on');
return;
end

% close
function close(self)
   close(self.handle);
return;
end

% io
function pgfprint(self,varargin);
   default_odir='./';
   default_frmt='eps';
   default_tbm='tbmb';
   default_auxfrmt='png';
   default_auxleg=true;
   default_README='*';
   default_panel='*';
   
   valid_odir = @(x) exist(x,'dir');
   valid_frmt = @(x) any(find(strcmp(x,{'eps','png','jpg','jpeg'})));
   
   p=inputParser();
   addOptional(p,'odir',   default_odir,valid_odir );
   addOptional(p,'frmt',   default_frmt,valid_frmt );
   addOptional(p,'auxfrmt',default_auxfrmt,valid_frmt );
   addOptional(p,'tbm'    ,default_tbm);
   parse(p,varargin{:});
   odir=p.Results.odir;
   frmt=p.Results.frmt;
   tbm=p.Results.tbm;
   auxfrmt=p.Results.auxfrmt;
   
   figlabel=self.label;
   if length(figlabel)>1
      ofnm_fig=sprintf("%s/fig_%d-%s_WITHAXES.%s",odir,self.handle.Number,figlabel,auxfrmt);
   else
      ofnm_fig=sprintf("%s/fig_%d_WITHAXES.%s",odir,self.handle.Number,auxfrmt);
   end
   saveas(self.handle,ofnm_fig,auxfrmt);

  
   if length(self.axlist) > 1
      error("I can only handle one axes");
   end
   xl=get(self.ax,'XLim'); yl=get(self.ax,'YLim');


   set(self.handle, 'Units', 'inches');
   set(self.ax    , 'Units', 'inches');

   figsize0= get(self.handle,'Position');
   axsize0 = get(self.ax,'Position');

   set(self.handle, 'Position', [1.0 1.0 self.ww self.hh]);
   set(self.ax    , 'Position', [0.0 0.0 self.ww self.hh]);

   figsize = get(self.handle,'Position');
   figunits = get(self.handle,'Units');
   papercolor=self.papercolor;

   set(self.handle,'Position', figsize);
   set(self.handle,'PaperPosition',[0 0 figsize(3) figsize(4)]);
   set(self.handle,'PaperUnits',figunits);
   set(self.handle,'PaperSize',[figsize(3) figsize(4)]);
   set(self.handle,'PaperPosition',[0 0 figsize(3) figsize(4)]);

   set(self.handle, 'InvertHardCopy','off');
   ax2 = axes('Parent',self.handle,'Units', 'normalize', ...
    'Position', [0 0 1 1], 'Color', papercolor,'XColor',papercolor,'YColor',papercolor);
   set(ax2, 'Position', [0 0 1 1]);
   uistack(ax2, 'bottom')


   if length(figlabel)>1
      ofnm_fig=sprintf("%s/fig_%d-%s.%s",odir,self.handle.Number,figlabel,frmt);
   else
      ofnm_fig=sprintf("%s/fig_%d.%s",odir,self.handle.Number,frmt);
   end
   axis(self.ax,'off');
   saveargs = {};
   if strcmp(frmt,'eps') 
      print(self.handle,ofnm_fig,'-depsc',saveargs{:});
   else
      error("frmt");
   end
   delete(ax2)
   axis(self.ax,'on');

   set(self.handle,'Position',figsize0);
   set(self.ax,'Position',axsize0);

   if length(figlabel)>1
      ofnm_tex=sprintf("%s/fig_%d-%s.tex",odir,self.handle.Number,figlabel);
   else
      ofnm_tex=sprintf("%s/fig_%d.tex",odir,self.handle.Number);
   end
   fid=fopen(ofnm_tex,'w');
   fprintf(fid,'%% width=%f in,\n',self.ww);
   fprintf(fid,'%% height=%f in,\n',self.hh);
   fprintf(fid,'%% xlabel={%f}\n',get(get(self.ax,'XLabel'),'String'));
   fprintf(fid,'%% ylabel={%f}\n',get(get(self.ax,'YLabel'),'String'));
   fprintf(fid,'%%\\begin{axis}[scale only axis, enlargelimits=false, axis on top,width=%fin,height=%fin]\n',self.ww,self.hh);
   if length(figlabel)>1
      fprintf(fid,'\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d-%s.%s};',xl(1),xl(2),yl(1),yl(2),tbm,self.handle.Number,figlabel,frmt);
   else
      fprintf(fid,'\\addplot[] graphics [xmin=%E,xmax=%E,ymin=%E,ymax=%E] {\\%s/fig_%d.%s};',xl(1),xl(2),yl(1),yl(2),tbm,self.handle.Number,frmt);
   end
   fclose(fid);
return;
end

function saveLine2D(self,varargin)
   default_odir='./';
   valid_odir = @(x) exist(x,'dir');
   p=inputParser();
   addOptional(p,'odir',   default_odir,valid_odir );
   parse(p,varargin{:});
   odir=p.Results.odir;

   figlabel=self.label;

   mychildren=findobj(self.ax,'Type','Line');
   nl=length(mychildren);

   if length(figlabel)>1
      odir_data=sprintf("%s/fig_%d-%s_data",odir,self.handle.Number,figlabel);
   else
      odir_data=sprintf("%s/fig_%d_data",odir,self.handle.Number);
   end
   if ~exist(odir_data,'dir'), system(sprintf('mkdir %s',odir_data));end
   for i1 = 1:nl
      l=mychildren(i1);
      tag=get(l,'Tag');
      if ~isempty(tag)
         ofnm=sprintf("%s/%s.dat",odir_data, strrep(tag,' ','_'));
         x=get(l,'XData');
         y=get(l,'YData');
         fid=fopen(ofnm,'w');
         nd=length(x(:));
         fprintf(fid,'%25s %25s\n','x','y');
         for i2 = 1:nd
            fprintf(fid,'%25.15E %25.15E\n',x(i2),y(i2));
         end
         fclose(fid);
      end
   end

return;
end


end

end


